<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>GenomicHilbertCurve: specific for genomic data • HilbertCurve</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="GenomicHilbertCurve: specific for genomic data">
<meta property="og:description" content="HilbertCurve">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">HilbertCurve</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.33.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Vignettes</li>
    <li>
      <a href="../articles/hc_general.html">Making 2D Hilbert Curve</a>
    </li>
    <li>
      <a href="../articles/hc_genome.html">GenomicHilbertCurve: specific for genomic data</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">More applications</li>
    <li>
      <a href="../articles/cpan.html">Visualize CPAN modules with Hilbert curve</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jokergoo/HilbertCurve/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>GenomicHilbertCurve: specific for genomic data</h1>
                        <h4 data-toc-skip class="author">Zuguang Gu (<a href="mailto:z.gu@dkfz.de" class="email">z.gu@dkfz.de</a>)</h4>
            
            <h4 data-toc-skip class="date">2024-02-27</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/jokergoo/HilbertCurve/blob/HEAD/vignettes/hc_genome.Rmd" class="external-link"><code>vignettes/hc_genome.Rmd</code></a></small>
      <div class="hidden name"><code>hc_genome.Rmd</code></div>

    </div>

    
    
<p>To make it easy with genomic data which is always stored as a data frame or a <code>GRanges</code> object, the package provides specific class and functions to deal with. <code><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve()</a></code> initializes the Hilbert curve by specifying chromosomes or specific background regions. After that, the data frame or <code>GRanges</code> object which contains genomic intervals can be mapped to the curve.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hc</span> <span class="op">=</span> <span class="fu"><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve</a></span><span class="op">(</span><span class="va">chr</span>, <span class="va">species</span>, <span class="va">...</span><span class="op">)</span></span>
<span><span class="va">hc</span> <span class="op">=</span> <span class="fu"><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve</a></span><span class="op">(</span><span class="va">background</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
<p>Following functions are depatched to the <code>GenomicHilbertCurve</code> class, and the second argument should be a data frame (which is in bed-format) or a <code>GRanges</code> object.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/hc_points-dispatch.html">hc_points</a></span><span class="op">(</span><span class="va">hc</span>, <span class="va">gr</span>, <span class="va">...</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/hc_rect-dispatch.html">hc_rect</a></span><span class="op">(</span><span class="va">hc</span>, <span class="va">gr</span>, <span class="va">...</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/hc_polygon-dispatch.html">hc_polygon</a></span><span class="op">(</span><span class="va">hc</span>, <span class="va">gr</span>, <span class="va">...</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/hc_segments-dispatch.html">hc_segments</a></span><span class="op">(</span><span class="va">hc</span>, <span class="va">gr</span>, <span class="va">...</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/hc_text-dispatch.html">hc_text</a></span><span class="op">(</span><span class="va">hc</span>, <span class="va">gr</span>, <span class="va">...</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span><span class="op">(</span><span class="va">hc</span>, <span class="va">gr</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
<p>Following illustrates several examples of making genome-level Hilbert curves.</p>
<div class="section level2">
<h2 id="genes">Genes<a class="anchor" aria-label="anchor" href="#genes"></a>
</h2>
<p>The first example visualizes genes on chromosome 1 (<a href="https://genome.ucsc.edu/cgi-bin/hgTables?hgsid=451688581_FoCuggvGvavyfGhdj9sSGrGlnErE&amp;clade=mammal&amp;org=Human&amp;db=hg38&amp;hgta_group=genes&amp;hgta_track=refGene&amp;hgta_table=0&amp;hgta_regionType=genome&amp;position=chr9%3A133252000-133280861&amp;hgta_outputType=primaryTable&amp;hgta_outFileName=" class="external-link">RefSeq genes for human, hg19</a>). Here random colors are used to distinguish different genes.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: package 'S4Vectors' was built under R version 4.3.2</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/HilbertCurve" class="external-link">HilbertCurve</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/circlize" class="external-link">circlize</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"refseq_chr1.RData"</span>, package <span class="op">=</span> <span class="st">"HilbertCurve"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">hc</span> <span class="op">=</span> <span class="fu"><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve</a></span><span class="op">(</span>chr <span class="op">=</span> <span class="st">"chr1"</span>, level <span class="op">=</span> <span class="fl">5</span>, reference <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>    reference_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>lty <span class="op">=</span> <span class="fl">1</span>, col <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span>, arrow <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/hc_segments-dispatch.html">hc_segments</a></span><span class="op">(</span><span class="va">hc</span>, <span class="va">g</span>, gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>lwd <span class="op">=</span> <span class="fl">6</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/rand_color.html" class="external-link">rand_color</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="hc_genome_files/figure-html/unnamed-chunk-4-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="sequence-conservation">Sequence conservation<a class="anchor" aria-label="anchor" href="#sequence-conservation"></a>
</h2>
<p>The conserved regions on human chromosome 1 between mouse and human or between zebrafish and human are downloaded from <a href="http://genome.ucsc.edu/cgi-bin/hgTables?hgsid=472144813_MQ4BbmgW37du53Nw6n9SKKy6mb6r&amp;clade=mammal&amp;org=Human&amp;db=hg38&amp;hgta_group=compGeno&amp;hgta_track=placentalChainNet&amp;hgta_table=netMm10&amp;hgta_regionType=genome&amp;position=chr9%3A133252000-133280861&amp;hgta_outputType=primaryTable&amp;hgta_outFileName=" class="external-link">UCSC Table Browser</a>. The parameters for downloading are:</p>
<pre><code>clade: Mammal
genome: Human
assembly: Feb. 2009(GRCh37/hg19)
group: Comparative Genomics
= for mouse
track: Placental Chain/Net
table: Mouse Net (netMm10)
= for zebrafish
track: Vertebrate Chain/Net
table: Zebrafish Net(netDanRer7)
</code></pre>
<p>The “net” alignment which is used here shows the best alignment (the alignment algorithm allows longer gaps than traditional affine gap scoring systems) between each part in human genome and other genomes. For a detailed description of the alignment strategy, please go to <a href="http://genome.ucsc.edu/cgi-bin/hgTables?hgsid=472144813_MQ4BbmgW37du53Nw6n9SKKy6mb6r&amp;clade=mammal&amp;org=Human&amp;db=hg38&amp;hgta_group=compGeno&amp;hgta_track=placentalChainNet&amp;hgta_table=netMm10&amp;hgta_regionType=genome&amp;position=chr9%3A133252000-133280861&amp;hgta_outputType=primaryTable&amp;hgta_outFileName=" class="external-link">UCSC Table Browser</a>, and click <strong>describe table schema</strong> button.</p>
<p>In following example, the conservation data is mapped to a level 6 Hilbert curve under “normal” mode. Points are used as the graphic and each segment on the Hilbert curve is split by 3 points. In this case, each point on the curve represents approximately 30kb. Only chromosome 1 is visualized.</p>
<p>The point is fully red if the window which is represented by this point is completely covered by the conserved regions, and it is yellow if it is not covered by any region. Colors are interpolated linearly between red and yellow if it is partially covered by the conserved regions.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># for generating the legend</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/ComplexHeatmap" class="external-link">ComplexHeatmap</a></span><span class="op">)</span></span>
<span><span class="va">lgd</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Legend.html" class="external-link">Legend</a></span><span class="op">(</span>col_fun <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html" class="external-link">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"yellow"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span>, </span>
<span>    title <span class="op">=</span> <span class="st">"Conservation"</span>,</span>
<span>    at <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.2</span>, <span class="fl">0.4</span>, <span class="fl">0.6</span>, <span class="fl">0.8</span>, <span class="fl">1</span><span class="op">)</span>, </span>
<span>    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0%"</span>, <span class="st">"20%"</span>, <span class="st">"40%"</span>, <span class="st">"60%"</span>, <span class="st">"80%"</span>, <span class="st">"100%"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">chr1_len</span> <span class="op">=</span> <span class="fl">249250621</span></span></code></pre></div>
<p>Conservation between human and mouse.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"mouse_net.RData"</span>, package <span class="op">=</span> <span class="st">"HilbertCurve"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">mouse</span><span class="op">)</span> <span class="op">=</span> <span class="va">chr1_len</span> <span class="co"># it is only used to extract the complement</span></span>
<span><span class="va">nonmouse</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/inter-range-methods.html" class="external-link">gaps</a></span><span class="op">(</span><span class="va">mouse</span><span class="op">)</span>; <span class="va">nonmouse</span> <span class="op">=</span> <span class="va">nonmouse</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/strand.html" class="external-link">strand</a></span><span class="op">(</span><span class="va">nonmouse</span><span class="op">)</span> <span class="op">==</span> <span class="st">"*"</span><span class="op">]</span></span>
<span><span class="va">gr</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">mouse</span>, <span class="va">nonmouse</span><span class="op">)</span></span>
<span><span class="va">col</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"red"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">mouse</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"yellow"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">nonmouse</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">hc</span> <span class="op">=</span> <span class="fu"><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve</a></span><span class="op">(</span>chr <span class="op">=</span> <span class="st">"chr1"</span>, level <span class="op">=</span> <span class="fl">6</span>, </span>
<span>    title <span class="op">=</span> <span class="st">"Conservation between mouse and human on chr1"</span>,</span>
<span>    legend <span class="op">=</span> <span class="va">lgd</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/hc_points-dispatch.html">hc_points</a></span><span class="op">(</span><span class="va">hc</span>, <span class="va">gr</span>, np <span class="op">=</span> <span class="fl">3</span>, gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>col <span class="op">=</span> <span class="cn">NA</span>, fill <span class="op">=</span> <span class="va">col</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="hc_genome_files/figure-html/unnamed-chunk-6-1.png" width="720" style="display: block; margin: auto;"></p>
<p>Conservation between human and zebrafish.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"zebrafish_net.RData"</span>, package <span class="op">=</span> <span class="st">"HilbertCurve"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlengths</a></span><span class="op">(</span><span class="va">zebrafish</span><span class="op">)</span> <span class="op">=</span> <span class="va">chr1_len</span></span>
<span><span class="va">nonzebrafish</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/inter-range-methods.html" class="external-link">gaps</a></span><span class="op">(</span><span class="va">zebrafish</span><span class="op">)</span>; <span class="va">nonzebrafish</span> <span class="op">=</span> <span class="va">nonzebrafish</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/strand.html" class="external-link">strand</a></span><span class="op">(</span><span class="va">nonzebrafish</span><span class="op">)</span> <span class="op">==</span> <span class="st">"*"</span><span class="op">]</span></span>
<span><span class="va">gr</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">zebrafish</span>, <span class="va">nonzebrafish</span><span class="op">)</span></span>
<span><span class="va">col</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"red"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">zebrafish</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"yellow"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">nonzebrafish</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">hc</span> <span class="op">=</span> <span class="fu"><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve</a></span><span class="op">(</span>chr <span class="op">=</span> <span class="st">"chr1"</span>, level <span class="op">=</span> <span class="fl">6</span>, </span>
<span>    title <span class="op">=</span> <span class="st">"Conservation between zebrafish and human on chr1"</span>,</span>
<span>    legend <span class="op">=</span> <span class="va">lgd</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/hc_points-dispatch.html">hc_points</a></span><span class="op">(</span><span class="va">hc</span>, <span class="va">gr</span>, np <span class="op">=</span> <span class="fl">3</span>, gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>col <span class="op">=</span> <span class="cn">NA</span>, fill <span class="op">=</span> <span class="va">col</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="hc_genome_files/figure-html/unnamed-chunk-7-1.png" width="720" style="display: block; margin: auto;"></p>
<p>Compareing these two plots, we have following conclusions:</p>
<ol style="list-style-type: decimal">
<li>Although it is already known that mouse is far more orthologically close to human, the Hilber curve illustrates more clearly that almost all parts of human chromosome 1 have similar sequences on mouse genome.</li>
<li>The start and end of the human chromosome 1 is less conserved to mouse compared to other parts in the chromosome. Also there are several less conserved regions inside the human chromosome.</li>
<li>It illustrates how the regions which are conserved to zebrafish spread in the human chromosome and it gives an estimation of the positions and length of the conserved regions.</li>
</ol>
<p>The idea of visualizing sequencing conservation through the Hilbert curve is from <a href="http://mkweb.bcgsc.ca/hilbert/scientificamerican.mhtml" class="external-link">here</a>.</p>
</div>
<div class="section level2">
<h2 id="gc-percent">GC percent<a class="anchor" aria-label="anchor" href="#gc-percent"></a>
</h2>
<p>Following figure visualizes GC percent on chromosome 1, under “normal” mode. The GC content per 1KB window is calculated by <code>hgGcPercent.bin</code> binary which is downloaded from <a href="http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/" class="external-link uri">http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/</a>. To speed up, we only used chromosome 1 for demonstration.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">pipe</a></span><span class="op">(</span><span class="st">"awk '$1==\"chr1\"' ~/HilbertCurveTest/hg19_gc_percent_window1000.bed"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">col_fun</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html" class="external-link">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">df</span><span class="op">[[</span><span class="fl">5</span><span class="op">]</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">0.5</span>, <span class="fl">0.9</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"green"</span>, <span class="st">"#FFFFCC"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">lgd</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Legend.html" class="external-link">Legend</a></span><span class="op">(</span>col_fun <span class="op">=</span> <span class="va">col_fun</span>, title <span class="op">=</span> <span class="st">"GC percent"</span>,</span>
<span>    at <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">300</span>, <span class="fl">400</span>, <span class="fl">500</span>, <span class="fl">600</span><span class="op">)</span>, </span>
<span>    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"30%"</span>, <span class="st">"40%"</span>, <span class="st">"50%"</span>, <span class="st">"60%"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">hc</span> <span class="op">=</span> <span class="fu"><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve</a></span><span class="op">(</span>chr <span class="op">=</span> <span class="st">"chr1"</span>, level <span class="op">=</span> <span class="fl">6</span>, legend <span class="op">=</span> <span class="va">lgd</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/hc_points-dispatch.html">hc_points</a></span><span class="op">(</span><span class="va">hc</span>, <span class="va">df</span>, np <span class="op">=</span> <span class="fl">3</span>, gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="fu">col_fun</span><span class="op">(</span><span class="va">df</span><span class="op">[[</span><span class="fl">5</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, col <span class="op">=</span> <span class="fu">col_fun</span><span class="op">(</span><span class="va">df</span><span class="op">[[</span><span class="fl">5</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/hc_rect-dispatch.html">hc_rect</a></span><span class="op">(</span><span class="va">hc</span>, <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/inter-range-methods.html" class="external-link">reduce</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span>, gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"#00000020"</span>, col <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>
</p>
<center>
<img src="gc_percent_chr1_points.png">
</center>

<p>Or visualize GC percent on chromosome 1, under “pixel” mode (you may see the difference between “normal” mode and “pixel” mode here):</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hc</span> <span class="op">=</span> <span class="fu"><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve</a></span><span class="op">(</span>chr <span class="op">=</span> <span class="st">"chr1"</span>, level <span class="op">=</span> <span class="fl">9</span>, mode <span class="op">=</span> <span class="st">"pixel"</span>, legend <span class="op">=</span> <span class="va">lgd</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span><span class="op">(</span><span class="va">hc</span>, <span class="va">df</span>, col <span class="op">=</span> <span class="fu">col_fun</span><span class="op">(</span><span class="va">df</span><span class="op">[[</span><span class="fl">5</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span><span class="op">(</span><span class="va">hc</span>, <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/inter-range-methods.html" class="external-link">reduce</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"#00000020"</span><span class="op">)</span></span></code></pre></div>
<p>
</p>
<center>
<img src="gc_percent_chr1.png">
</center>

<p>Background is not necessary to be the whole chromosome, it can also be self-defined regions. The background can be specified as a <code>GRanges</code> object and pass to <code>background</code> argument. In following example, only the first half part of chromosome 1 is used to initialize the curve.</p>
<p><strong>Note: when using <code>background</code>, one chromosome can only exist in one row in <code>background</code>.</strong></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">background</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span>seqnames <span class="op">=</span> <span class="st">"chr1"</span>, ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">ceiling</a></span><span class="op">(</span><span class="va">chr1_len</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">hc</span> <span class="op">=</span> <span class="fu"><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve</a></span><span class="op">(</span>background <span class="op">=</span> <span class="va">background</span>, level <span class="op">=</span> <span class="fl">9</span>, mode <span class="op">=</span> <span class="st">"pixel"</span>, legend <span class="op">=</span> <span class="va">lgd</span>,</span>
<span>    title <span class="op">=</span> <span class="st">"First half of chromosome 1"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span><span class="op">(</span><span class="va">hc</span>, <span class="va">df</span>, col <span class="op">=</span> <span class="fu">col_fun</span><span class="op">(</span><span class="va">df</span><span class="op">[[</span><span class="fl">5</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span><span class="op">(</span><span class="va">hc</span>, <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/inter-range-methods.html" class="external-link">reduce</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"#00000020"</span><span class="op">)</span></span></code></pre></div>
<p>
</p>
<center>
<img src="gc_percent_half_chr1.png">
</center>

</div>
<div class="section level2">
<h2 id="histone-modifications">Histone modifications<a class="anchor" aria-label="anchor" href="#histone-modifications"></a>
</h2>
<p>Different modifications (e.g. methylation, acetylation) of the histones affect the chromatin-DNA interaction, thus play roles in transcription regulation. Different types of modifications relates to different types of gene regulation. E.g. H3H4me3 modification is found in actively transcribed promoters. ChIP-sequencing (ChIPSeq) utilizes highthroughput DNA sequencing technology to identify status of chromatin states genome-wide, thus it provides a way to study gene regulation in a global aspect.</p>
<p>In this section, we visualize four types of histone modifications (H3K27ac, H3K36me3, H3K4me3, H3K9me3). The ChIPSeq data are downloaded from Roadmap (<a href="http://genboree.org/EdaccData/Release-9/sample-experiment/Lung/" class="external-link uri">http://genboree.org/EdaccData/Release-9/sample-experiment/Lung/</a>, The bed files are alignments of reads, you need to convert to the basepair or window-based coverage before using following R commands). Following files for four histone modifications are used to visualize:</p>
<pre><code><span></span>
<span><span class="va">UCSD.Lung.H3K27ac.STL002.bed</span></span>
<span><span class="va">UCSD.Lung.H3K36me3.STL002.bed</span></span>
<span><span class="va">UCSD.Lung.H3K4me3.STL002.bed</span></span>
<span><span class="va">UCSD.Lung.H3K9me3.STL002.bed</span></span></code></pre>
<p>Since we are going to plot more than one histome marks, the code for plotting are wrapped as a simple function:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/GetoptLong" class="external-link">GetoptLong</a></span><span class="op">)</span></span>
<span><span class="va">plot_histone_mark</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">mark</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">df</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">pipe</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GetoptLong/man/qq.html" class="external-link">qq</a></span><span class="op">(</span><span class="st">"awk '$5&gt;0 &amp;&amp; $1==\"chr1\"' ~/HilbertCurveTest/UCSD.Lung.@{mark}.STL002.bed"</span><span class="op">)</span><span class="op">)</span>, </span>
<span>        sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span></span>
<span>    <span class="va">col_fun</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html" class="external-link">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">df</span><span class="op">[[</span><span class="fl">5</span><span class="op">]</span><span class="op">]</span>, <span class="fl">0.99</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">lgd1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Legend.html" class="external-link">Legend</a></span><span class="op">(</span>col_fun <span class="op">=</span> <span class="va">col_fun</span>, title <span class="op">=</span> <span class="st">"Intensity"</span><span class="op">)</span></span>
<span>    <span class="va">lgd2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Legend.html" class="external-link">Legend</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">mark</span>, <span class="st">"gene"</span><span class="op">)</span>, legend_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#FF0000"</span>, <span class="st">"#CCCCCC"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        title <span class="op">=</span> <span class="st">"Layer"</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">hc</span> <span class="op">=</span> <span class="fu"><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve</a></span><span class="op">(</span>chr <span class="op">=</span> <span class="st">"chr1"</span>, level <span class="op">=</span> <span class="fl">9</span>, mode <span class="op">=</span> <span class="st">"pixel"</span>, </span>
<span>        title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GetoptLong/man/qq.html" class="external-link">qq</a></span><span class="op">(</span><span class="st">"Intensity of @{mark} mark on chr1"</span><span class="op">)</span>, legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">lgd1</span>, <span class="va">lgd2</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span><span class="op">(</span><span class="va">hc</span>, <span class="va">df</span>, col <span class="op">=</span> <span class="fu">col_fun</span><span class="op">(</span><span class="va">df</span><span class="op">[[</span><span class="fl">5</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span><span class="op">(</span><span class="va">hc</span>, <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/inter-range-methods.html" class="external-link">reduce</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"#00000010"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>In above code, intensities of histone modifications are mapped to a level 9 Hilbert curve so that each pixel approximately represents a window of 950bp. In each plot, there is an additional grey layer which represents the regions of genes (RefSeq genes) (<code>g</code> is from <code>system.file("extdata", "refseq_chr1.RData", package = "HilbertCurve")</code>).</p>
<p>H3K27ac is a mark for active enhancers which is narrow and sharp. The Hilbert curve shows this attributes and also shows enhancers may exist in form of clusters.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_histone_mark</span><span class="op">(</span><span class="st">"H3K27ac"</span><span class="op">)</span></span></code></pre></div>
<p>
</p>
<center>
<img src="H3K27ac_chr1.png">
</center>

<p>H3K36me3 is a mark for actively transcribed gene bodies. The Hilbert curve shows that basically most of H3K36me3 modification with high intensity exists as big blocks.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_histone_mark</span><span class="op">(</span><span class="st">"H3K36me3"</span><span class="op">)</span></span></code></pre></div>
<p>
</p>
<center>
<img src="H3K36me3_chr1.png">
</center>

<p>H3K4me3 is a mark for TSS of actively transcribed genes which is also narrow and sharp. The Hilbert curve confirms this and also shows the size of marks are quite consistent.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_histone_mark</span><span class="op">(</span><span class="st">"H3K4me3"</span><span class="op">)</span></span></code></pre></div>
<p>
</p>
<center>
<img src="H3K4me3_chr1.png">
</center>

<p>H3K9me3 is a mark for repressed genes. The Hilbert curve shows that H3K9me3 modications spread over huge regions in the chromosome.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_histone_mark</span><span class="op">(</span><span class="st">"H3K9me3"</span><span class="op">)</span></span></code></pre></div>
<p>
</p>
<center>
<img src="H3K9me3_chr1.png">
</center>

<p>When there are more than one layers added and users want to see the correspondance between different genomic features, they can define a color overlay function to enhance the visual effects of the overlapping areas. In following example, when regions for H3K36me3 mark are overlapped with gene bodies, the color theme is changed to white-purple.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>df <span class="ot">=</span> <span class="fu">read.table</span>(<span class="fu">pipe</span>(<span class="st">"awk '$5&gt;0 &amp;&amp; $1==</span><span class="sc">\"</span><span class="st">chr1</span><span class="sc">\"</span><span class="st">' ~/HilbertCurveTest/UCSD.Lung.H3K36me3.STL002.bed"</span>), </span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>col_fun <span class="ot">=</span> <span class="fu">colorRamp2</span>(<span class="fu">c</span>(<span class="dv">0</span>, <span class="fu">quantile</span>(df[[<span class="dv">5</span>]], <span class="fl">0.99</span>)), <span class="fu">c</span>(<span class="st">"white"</span>, <span class="st">"red"</span>))</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>col_fun_new <span class="ot">=</span> <span class="fu">colorRamp2</span>(<span class="fu">c</span>(<span class="dv">0</span>, <span class="fu">quantile</span>(df[[<span class="dv">5</span>]], <span class="fl">0.99</span>)), <span class="fu">c</span>(<span class="st">"white"</span>, <span class="st">"purple"</span>))</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>lgd1 <span class="ot">=</span> <span class="fu">Legend</span>(<span class="at">col_fun =</span> col_fun, <span class="at">title =</span> <span class="st">"Intensity"</span>)</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>lgd2 <span class="ot">=</span> <span class="fu">Legend</span>(<span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"H3K36me3"</span>, <span class="st">"overlap"</span>, <span class="st">"gene"</span>),</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend_gp =</span> <span class="fu">gpar</span>(<span class="at">fill =</span> <span class="fu">c</span>(<span class="st">"#FF0000"</span>, <span class="st">"purple"</span>, <span class="st">"#CCCCCC"</span>)), <span class="at">title =</span> <span class="st">"Layer"</span>)</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="st">"H3K36me3_chr1_overlap.png"</span>, <span class="at">width =</span> <span class="dv">500</span>, <span class="at">height =</span> <span class="dv">500</span>)</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>    title <span class="ot">=</span> <span class="st">"Intensity of H3K36me3 mark on chr1"</span>, legend <span class="ot">=</span> <span class="fu">list</span>(lgd1, lgd2)<span class="er">)</span></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a><span class="fu">hc_layer</span>(hc, df, <span class="at">col =</span> <span class="fu">col_fun</span>(df[[<span class="dv">5</span>]]))</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a><span class="fu">hc_layer</span>(hc, <span class="fu">reduce</span>(g), <span class="at">col =</span> <span class="st">"#00000010"</span>,</span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">overlay =</span> <span class="cf">function</span>(r0, g0, b0, r, g, b, alpha) {</span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a>        l <span class="ot">=</span> <span class="sc">!</span><span class="fu">is_white</span>(r0, g0, b0)</span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a>        v <span class="ot">=</span> <span class="fu">col2value</span>(r0[l], g0[l], b0[l], <span class="at">col_fun =</span> col_fun)</span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a>        col_new <span class="ot">=</span> <span class="fu">col_fun_new</span>(v, <span class="at">return_rgb =</span> <span class="cn">TRUE</span>)</span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a>        r0[l] <span class="ot">=</span> col_new[, <span class="dv">1</span>]</span>
<span id="cb17-19"><a href="#cb17-19" aria-hidden="true" tabindex="-1"></a>        g0[l] <span class="ot">=</span> col_new[, <span class="dv">2</span>]</span>
<span id="cb17-20"><a href="#cb17-20" aria-hidden="true" tabindex="-1"></a>        b0[l] <span class="ot">=</span> col_new[, <span class="dv">3</span>]</span>
<span id="cb17-21"><a href="#cb17-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-22"><a href="#cb17-22" aria-hidden="true" tabindex="-1"></a>        <span class="fu">default_overlay</span>(r0, g0, b0, r, g, b, alpha)</span>
<span id="cb17-23"><a href="#cb17-23" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb17-24"><a href="#cb17-24" aria-hidden="true" tabindex="-1"></a><span class="fu">invisible</span>(<span class="fu">dev.off</span>())</span></code></pre></div>
<p>
</p>
<center>
<img src="H3K36me3_chr1_overlap.png">
</center>

</div>
<div class="section level2">
<h2 id="methylation">Methylation<a class="anchor" aria-label="anchor" href="#methylation"></a>
</h2>
<p>Next figure visualizs methylation on human chromosome 1. Blue corresponds to un-methylation and red corresponds to full-methylation. One thing which is very important here is that methylation is always measured at single CpG sites, so the mode for averaging should be set to <code>absolute</code> in order to sumarize mean methylation values only on CpG sites. Data are from <a href="http://genboree.org/EdaccData/Release-9/sample-experiment/Lung/" class="external-link uri">http://genboree.org/EdaccData/Release-9/sample-experiment/Lung/</a>.</p>
<p>The methylation plot shows quite clear patterns for un-methylated regions which are small and sharp, fully methylated regions and intermediate methylated regions which are represented as large blocks.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">pipe</a></span><span class="op">(</span><span class="st">"awk '$1==\"chr1\"' ~/HilbertCurveTest/UCSD.Lung.Bisulfite-Seq.STL002.bed"</span><span class="op">)</span>, </span>
<span>    sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span></span>
<span><span class="va">col_fun</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html" class="external-link">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.5</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"white"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">lgd</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Legend.html" class="external-link">Legend</a></span><span class="op">(</span>col_fun <span class="op">=</span> <span class="va">col_fun</span>, title <span class="op">=</span> <span class="st">"Methylation"</span><span class="op">)</span></span>
<span><span class="va">hc</span> <span class="op">=</span> <span class="fu"><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve</a></span><span class="op">(</span>chr <span class="op">=</span> <span class="st">"chr1"</span>, level <span class="op">=</span> <span class="fl">9</span>, mode <span class="op">=</span> <span class="st">"pixel"</span>, legend <span class="op">=</span> <span class="va">lgd</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span><span class="op">(</span><span class="va">hc</span>, <span class="va">df</span>, col <span class="op">=</span> <span class="fu">col_fun</span><span class="op">(</span><span class="va">df</span><span class="op">[[</span><span class="fl">5</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, mean_mode <span class="op">=</span> <span class="st">"absolute"</span><span class="op">)</span></span></code></pre></div>
<p>
</p>
<center>
<img src="methylation_chr1.png">
</center>

</div>
<div class="section level2">
<h2 id="put-multiple-chromosomes-in-one-curve">Put multiple chromosomes in one curve<a class="anchor" aria-label="anchor" href="#put-multiple-chromosomes-in-one-curve"></a>
</h2>
<p>We demonstrated how to visualize one single chromosome through Hilbert curve, but sometimes we also want to put all chromosomes into one plot to get a full view of the genome. One solution is to construct a ‘huge fake’ chromosome which merges all real chromosomes. <code><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve()</a></code> makes it easy. You can specify multiple chromosomes by sending a vector to <code>chr</code> argument or specify more than one background intervals in <code>background</code> argument.</p>
<p>In following examples, copy number loss and gain for 22 human chromosomes are visualized. Also color for the overlapping regions between layer for gain events and layer for loss events is changed to purple by defining <code>overlay</code> function. Data is from <a href="http://dgv.tcag.ca/dgv/app/downloads" class="external-link">Database of Genomic Variants, CNV map</a>.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hc</span> <span class="op">=</span> <span class="fu"><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve</a></span><span class="op">(</span>chr <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span>, level <span class="op">=</span> <span class="fl">9</span>, mode <span class="op">=</span> <span class="st">"pixel"</span><span class="op">)</span></span>
<span><span class="va">df_gain</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"~/HilbertCurveTest/Stringent.Gain.hg19.2015-02-03.txt"</span>, </span>
<span>    header <span class="op">=</span> <span class="cn">TRUE</span>, stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span><span class="op">(</span><span class="va">hc</span>, <span class="va">df_gain</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span>
<span><span class="va">df_loss</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"~/HilbertCurveTest/Stringent.Loss.hg19.2015-02-03.txt"</span>, </span>
<span>    header <span class="op">=</span> <span class="cn">TRUE</span>, stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span><span class="op">(</span><span class="va">hc</span>, <span class="va">df_loss</span>, col <span class="op">=</span> <span class="st">"green"</span>, grid_line <span class="op">=</span> <span class="fl">3</span>, grid_line_col <span class="op">=</span> <span class="st">"grey"</span>,</span>
<span>    overlay <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">r0</span>, <span class="va">g0</span>, <span class="va">b0</span>, <span class="va">r</span>, <span class="va">g</span>, <span class="va">b</span>, <span class="va">alpha</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">l</span> <span class="op">=</span> <span class="op">!</span><span class="fu"><a href="../reference/is_white.html">is_white</a></span><span class="op">(</span><span class="va">r0</span>, <span class="va">g0</span>, <span class="va">b0</span><span class="op">)</span></span>
<span>        <span class="va">r</span><span class="op">[</span><span class="va">l</span><span class="op">]</span> <span class="op">=</span> <span class="fl">160</span><span class="op">/</span><span class="fl">255</span></span>
<span>        <span class="va">g</span><span class="op">[</span><span class="va">l</span><span class="op">]</span> <span class="op">=</span> <span class="fl">32</span><span class="op">/</span><span class="fl">255</span></span>
<span>        <span class="va">b</span><span class="op">[</span><span class="va">l</span><span class="op">]</span> <span class="op">=</span> <span class="fl">240</span><span class="op">/</span><span class="fl">255</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">r</span>, <span class="va">g</span>, <span class="va">b</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>
</p>
<center>
<img src="cnv_all_chromosomes.png">
</center>

<p>A map which shows the chromosome positions in Hilbert curve is very necessary for knowing which part in the curve represents which chromosome.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/hc_map-GenomicHilbertCurve-method.html">hc_map</a></span><span class="op">(</span><span class="va">hc</span><span class="op">)</span></span></code></pre></div>
<p>
</p>
<center>
<img src="map_all_chromosomes.png">
</center>

<p>The map can be overlayed to the Hilbert curve for easy correspondance.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hc</span> <span class="op">=</span> <span class="fu"><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve</a></span><span class="op">(</span>chr <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span>, level <span class="op">=</span> <span class="fl">9</span>, mode <span class="op">=</span> <span class="st">"pixel"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span><span class="op">(</span><span class="va">hc</span>, <span class="va">df_gain</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span><span class="op">(</span><span class="va">hc</span>, <span class="va">df_loss</span>, col <span class="op">=</span> <span class="st">"green"</span>, </span>
<span>    overlay <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">r0</span>, <span class="va">g0</span>, <span class="va">b0</span>, <span class="va">r</span>, <span class="va">g</span>, <span class="va">b</span>, <span class="va">alpha</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">l</span> <span class="op">=</span> <span class="op">!</span><span class="fu"><a href="../reference/is_white.html">is_white</a></span><span class="op">(</span><span class="va">r0</span>, <span class="va">g0</span>, <span class="va">b0</span><span class="op">)</span></span>
<span>        <span class="va">r</span><span class="op">[</span><span class="va">l</span><span class="op">]</span> <span class="op">=</span> <span class="fl">160</span><span class="op">/</span><span class="fl">255</span></span>
<span>        <span class="va">g</span><span class="op">[</span><span class="va">l</span><span class="op">]</span> <span class="op">=</span> <span class="fl">32</span><span class="op">/</span><span class="fl">255</span></span>
<span>        <span class="va">b</span><span class="op">[</span><span class="va">l</span><span class="op">]</span> <span class="op">=</span> <span class="fl">240</span><span class="op">/</span><span class="fl">255</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">r</span>, <span class="va">g</span>, <span class="va">b</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/hc_map-GenomicHilbertCurve-method.html">hc_map</a></span><span class="op">(</span><span class="va">hc</span>, add <span class="op">=</span> <span class="cn">TRUE</span>, fill <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/rand_color.html" class="external-link">rand_color</a></span><span class="op">(</span><span class="fl">22</span>, transparency <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>
</p>
<center>
<img src="cnv_all_chromosomes_overlay.png">
</center>

<p>If colors of chromosomes defuse the colors in the original Hilbert curve, you can only draw borders of chromosomes.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hc</span> <span class="op">=</span> <span class="fu"><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve</a></span><span class="op">(</span>chr <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span>, level <span class="op">=</span> <span class="fl">9</span>, mode <span class="op">=</span> <span class="st">"pixel"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span><span class="op">(</span><span class="va">hc</span>, <span class="va">df_gain</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span><span class="op">(</span><span class="va">hc</span>, <span class="va">df_loss</span>, col <span class="op">=</span> <span class="st">"green"</span>,</span>
<span>    overlay <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">r0</span>, <span class="va">g0</span>, <span class="va">b0</span>, <span class="va">r</span>, <span class="va">g</span>, <span class="va">b</span>, <span class="va">alpha</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">l</span> <span class="op">=</span> <span class="op">!</span><span class="fu"><a href="../reference/is_white.html">is_white</a></span><span class="op">(</span><span class="va">r0</span>, <span class="va">g0</span>, <span class="va">b0</span><span class="op">)</span></span>
<span>        <span class="va">r</span><span class="op">[</span><span class="va">l</span><span class="op">]</span> <span class="op">=</span> <span class="fl">160</span><span class="op">/</span><span class="fl">255</span></span>
<span>        <span class="va">g</span><span class="op">[</span><span class="va">l</span><span class="op">]</span> <span class="op">=</span> <span class="fl">32</span><span class="op">/</span><span class="fl">255</span></span>
<span>        <span class="va">b</span><span class="op">[</span><span class="va">l</span><span class="op">]</span> <span class="op">=</span> <span class="fl">240</span><span class="op">/</span><span class="fl">255</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">r</span>, <span class="va">g</span>, <span class="va">b</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/hc_map-GenomicHilbertCurve-method.html">hc_map</a></span><span class="op">(</span><span class="va">hc</span>, add <span class="op">=</span> <span class="cn">TRUE</span>, fill <span class="op">=</span> <span class="cn">NA</span>, border <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span></span></code></pre></div>
<p>
</p>
<center>
<img src="cnv_all_chromosomes_border.png">
</center>

<p>If you want to compare several plots simutaneously, you can use <code><a href="https://rdrr.io/r/grid/grid.layout.html" class="external-link">grid.layout()</a></code> to put them together. Remember to set <code>newpage</code> to <code>FALSE</code> in <code><a href="../reference/HilbertCurve.html">HilbertCurve()</a></code> or <code><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve()</a></code>. Also note we set <code>labels</code> in <code><a href="../reference/hc_map-GenomicHilbertCurve-method.html">hc_map()</a></code> to define the labels added to each chromosome.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot_curve</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">column</span>, <span class="va">...</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">cm</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/ColorMapping.html" class="external-link">ColorMapping</a></span><span class="op">(</span>levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gain"</span>, <span class="st">"loss"</span>, <span class="st">"both"</span><span class="op">)</span>, color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"red"</span>, <span class="st">"green"</span>, <span class="st">"purple"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">hc</span> <span class="op">=</span> <span class="fu"><a href="../reference/GenomicHilbertCurve.html">GenomicHilbertCurve</a></span><span class="op">(</span>chr <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span>, level <span class="op">=</span> <span class="fl">9</span>, mode <span class="op">=</span> <span class="st">"pixel"</span>,</span>
<span>        title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GetoptLong/man/qq.html" class="external-link">qq</a></span><span class="op">(</span><span class="st">"CNV for @{column}"</span><span class="op">)</span>, <span class="va">...</span><span class="op">)</span>  <span class="co">## `qq()` is from the GetoptLong package</span></span>
<span></span>
<span>    <span class="va">df</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"~/HilbertCurveTest/Stringent.Gain.hg19.2015-02-03.txt"</span>, header <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>        stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>    <span class="va">df</span> <span class="op">=</span> <span class="va">df</span><span class="op">[</span><span class="va">df</span><span class="op">[[</span><span class="va">column</span><span class="op">]</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">0</span>, , drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span>
<span>    <span class="fu"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span><span class="op">(</span><span class="va">hc</span>, <span class="va">df</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">df</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"~/HilbertCurveTest/Stringent.Loss.hg19.2015-02-03.txt"</span>, header <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>        stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>    <span class="va">df</span> <span class="op">=</span> <span class="va">df</span><span class="op">[</span><span class="va">df</span><span class="op">[[</span><span class="va">column</span><span class="op">]</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">0</span>, , drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span>
<span>    <span class="fu"><a href="../reference/hc_layer-dispatch.html">hc_layer</a></span><span class="op">(</span><span class="va">hc</span>, <span class="va">df</span>, col <span class="op">=</span> <span class="st">"green"</span>, </span>
<span>        overlay <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">r0</span>, <span class="va">g0</span>, <span class="va">b0</span>, <span class="va">r</span>, <span class="va">g</span>, <span class="va">b</span>, <span class="va">alpha</span><span class="op">)</span> <span class="op">{</span></span>
<span>            <span class="va">l</span> <span class="op">=</span> <span class="op">!</span><span class="fu"><a href="../reference/is_white.html">is_white</a></span><span class="op">(</span><span class="va">r0</span>, <span class="va">g0</span>, <span class="va">b0</span><span class="op">)</span></span>
<span>            <span class="va">r</span><span class="op">[</span><span class="va">l</span><span class="op">]</span> <span class="op">=</span> <span class="fl">160</span><span class="op">/</span><span class="fl">255</span></span>
<span>            <span class="va">g</span><span class="op">[</span><span class="va">l</span><span class="op">]</span> <span class="op">=</span> <span class="fl">32</span><span class="op">/</span><span class="fl">255</span></span>
<span>            <span class="va">b</span><span class="op">[</span><span class="va">l</span><span class="op">]</span> <span class="op">=</span> <span class="fl">240</span><span class="op">/</span><span class="fl">255</span></span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">r</span>, <span class="va">g</span>, <span class="va">b</span><span class="op">)</span></span>
<span>    <span class="op">}</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="../reference/hc_map-GenomicHilbertCurve-method.html">hc_map</a></span><span class="op">(</span><span class="va">hc</span>, labels <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">22</span>, add <span class="op">=</span> <span class="cn">TRUE</span>, fill <span class="op">=</span> <span class="cn">NA</span>, border <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">pn</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"African"</span>, <span class="st">"Asian"</span>, <span class="st">"European"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grid/viewports.html" class="external-link">pushViewport</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/grid/viewport.html" class="external-link">viewport</a></span><span class="op">(</span>layout <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/grid.layout.html" class="external-link">grid.layout</a></span><span class="op">(</span>nrow <span class="op">=</span> <span class="fl">1</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">pn</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grid/viewports.html" class="external-link">pushViewport</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/grid/viewport.html" class="external-link">viewport</a></span><span class="op">(</span>layout.pos.row <span class="op">=</span> <span class="fl">1</span>, layout.pos.col <span class="op">=</span> <span class="va">i</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu">plot_curve</span><span class="op">(</span><span class="va">pn</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, newpage <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/grid/viewports.html" class="external-link">upViewport</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>
</p>
<center>
<img src="cnv_compare.png">
</center>

</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">## Platform: x86_64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Ventura 13.2.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] C/UTF-8/C/C/C/C</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Europe/Berlin</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    grid      stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] GetoptLong_1.0.5      ComplexHeatmap_2.18.0 circlize_0.4.15       GenomicRanges_1.52.1 </span></span>
<span><span class="co">##  [5] GenomeInfoDb_1.36.4   IRanges_2.36.0        S4Vectors_0.40.2      BiocGenerics_0.48.1  </span></span>
<span><span class="co">##  [9] HilbertCurve_1.33.1   knitr_1.44            markdown_1.10        </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] shape_1.4.6             rjson_0.2.21            xfun_0.40               bslib_0.6.1            </span></span>
<span><span class="co">##  [5] GlobalOptions_0.1.2     lattice_0.21-9          vctrs_0.6.4             HilbertVis_1.58.0      </span></span>
<span><span class="co">##  [9] tools_4.3.1             bitops_1.0-7            parallel_4.3.1          cluster_2.1.4          </span></span>
<span><span class="co">## [13] RColorBrewer_1.1-3      polylabelr_0.2.0        desc_1.4.2              lifecycle_1.0.4        </span></span>
<span><span class="co">## [17] GenomeInfoDbData_1.2.10 compiler_4.3.1          stringr_1.5.0           textshaping_0.3.7      </span></span>
<span><span class="co">## [21] codetools_0.2-19        clue_0.3-65             htmltools_0.5.7         sass_0.4.8             </span></span>
<span><span class="co">## [25] RCurl_1.98-1.12         yaml_2.3.7              pkgdown_2.0.7           crayon_1.5.2           </span></span>
<span><span class="co">## [29] jquerylib_0.1.4         cachem_1.0.8            iterators_1.0.14        foreach_1.5.2          </span></span>
<span><span class="co">## [33] digest_0.6.33           stringi_1.7.12          purrr_1.0.2             rprojroot_2.0.3        </span></span>
<span><span class="co">## [37] fastmap_1.1.1           colorspace_2.1-0        cli_3.6.2               magrittr_2.0.3         </span></span>
<span><span class="co">## [41] rmarkdown_2.25          XVector_0.40.0          matrixStats_1.2.0       ragg_1.2.6             </span></span>
<span><span class="co">## [45] png_0.1-8               memoise_2.0.1           evaluate_0.22           doParallel_1.0.17      </span></span>
<span><span class="co">## [49] rlang_1.1.2             Rcpp_1.0.11             glue_1.6.2              jsonlite_1.8.8         </span></span>
<span><span class="co">## [53] R6_2.5.1                systemfonts_1.0.5       fs_1.6.3                zlibbioc_1.46.0</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Zuguang Gu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  <style>nav[data-toggle='toc'] .nav .nav {display: block;}</style>
</body>
</html>
